src/LanguageServer/Protocol/Workspaces/LspMiscellaneousFilesWorkspaceProvider.cs

```diff
@@ -3,27 +3,130 @@
 using Microsoft.CodeAnalysis.Host;
-using Microsoft.CodeAnalysis.Host.Mef;
-using Microsoft.CodeAnalysis.LanguageServer.Handler;
 using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
 using Microsoft.CommonLanguageServerProtocol.Framework;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.LanguageServer;
 
 /// <summary>
-/// Service to create <see cref="LspWorkspaceManager"/> instances.
-/// This is not exported as a <see cref="ILspServiceFactory"/> as it requires
-/// special base language server dependencies such as the <see cref="HostServices"/>
+/// Defines a default workspace for opened LSP files that are not found in any
+/// workspace registered by the <see cref="LspWorkspaceRegistrationService"/>.
+/// If a document added here is subsequently found in a registered workspace, 
+/// the document is removed from this workspace.
+/// 
+/// Future work for this workspace includes supporting basic metadata references (mscorlib, System dlls, etc),
+/// but that is dependent on having a x-plat mechanism for retrieving those references from the framework / sdk.
 /// </summary>
-[ExportCSharpVisualBasicStatelessLspService(typeof(LspMiscellaneousFilesWorkspaceProvider)), Shared]
-[method: ImportingConstructor]
-[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-internal sealed class LspMiscellaneousFilesWorkspaceProvider(IMetadataAsSourceFileService metadataAsSourceFileService) : ILspService
+internal sealed class LspMiscellaneousFilesWorkspaceProvider(ILspServices lspServices, IMetadataAsSourceFileService metadataAsSourceFileService, HostServices hostServices)
+    : Workspace(hostServices, WorkspaceKind.MiscellaneousFiles), ILspMiscellaneousFilesWorkspaceProvider, ILspWorkspace
 {
-    public LspMiscellaneousFilesWorkspace CreateLspMiscellaneousFilesWorkspace(ILspServices lspServices, HostServices hostServices)
+    public bool SupportsMutation => true;
+
+    // For this implementation, it's easiest to just have it inherit from Workspace, so we'll just return this.
+    public Workspace Workspace => this;
+
+    /// <summary>
+    /// Takes in a file URI and text and creates a misc project and document for the file.
+    /// 
+    /// Calls to this method and <see cref="TryRemoveMiscellaneousDocument(Uri, bool)"/> are made
+    /// from LSP text sync request handling which do not run concurrently.
+    /// </summary>
+    public TextDocument? AddMiscellaneousDocument(Uri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+
+        var container = new StaticSourceTextContainer(documentText);
+        if (metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, container, out var documentId))
+        {
+            var metadataWorkspace = metadataAsSourceFileService.TryGetWorkspace();
```

Comment:
maybe have the TryAddDocToWorkspace return the worskpace as well?  to avoid this?